crates/ruff_linter/src/rules/ruff/rules/if_key_in_dict_del.rs

```diff
@@ -0,0 +1,164 @@
+use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{
+    CmpOp, Expr, ExprCompare, ExprName, ExprSubscript, Stmt, StmtDelete, StmtIf,
+};
+use ruff_python_semantic::analyze::typing;
+use ruff_python_semantic::SemanticModel;
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+// Real type: Expr::Name;
+type Key = Expr;
+type Dict = Expr;
+
+/// ## What it does
+/// Checks for `if key in dictionary: del dictionary[key]`.
+///
+/// ## Why is this bad?
+/// Checking for the key's presence and deleting it
+/// requires calculating the hash twice.
```

Comment:
I think this gives the impression that `pop` is more performant, but that's not necessarily true. Here's a test stolen from [this stackoverflow comment](https://stackoverflow.com/questions/5713218/what-is-difference-between-pop-vs-del-for-deleting-a-dictionary-item-python#comment91857000_5713303) but run on Python 3.12:

```console
Python 3.12.7 (main, Oct 16 2024, 07:12:08) [Clang 18.1.8 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import timeit
>>> timeit.timeit(stmt="if 'two' in a: del a['two']", setup="a={'one': 1, 'two': 2, 'three': 3}", number=1000000)
0.023885874994448386
>>> timeit.timeit(stmt="a.pop('two', None)", setup="a={'one': 1, 'two': 2, 'three': 3}", number=1000000)
0.034923375002108514
```
So I think this is more of a style/readability lint than a performance lint (but I still like the lint!)

Reply:
I agree that this rule primarily suggests simplifying (reducing complexity). It's less about the performance itself (although that falls out of it). 

That's why I would also write the motivation this way. I'm also not good at writing descriptions. That's why I often search for similar rules to get inspired (copy paste :P). In this case, the flake8-simplify rules are a good place to start https://docs.astral.sh/ruff/rules/#flake8-simplify-sim